cpp/include/raft/core/resource/device_memory_resource.hpp

```diff
@@ -41,36 +66,109 @@ class device_memory_resource : public resource {
  */
 class workspace_resource_factory : public resource_factory {
  public:
-  workspace_resource_factory(rmm::mr::device_memory_resource* mr_ = nullptr) : mr(mr_) {}
-  resource_type get_resource_type() override { return resource_type::WORKSPACE_RESOURCE; }
-  resource* make_resource() override { return new device_memory_resource(mr); }
+  explicit workspace_resource_factory(
+    std::shared_ptr<rmm::mr::device_memory_resource> mr = {nullptr},
+    std::optional<std::size_t> allocation_limit         = std::nullopt,
+    std::optional<std::size_t> alignment                = std::nullopt)
+    : allocation_limit_(allocation_limit.value_or(default_allocation_limit())),
+      alignment_(alignment),
+      mr_(mr ? mr : default_memory_resource(allocation_limit_))
+  {
+  }
+
+  auto get_resource_type() -> resource_type override { return resource_type::WORKSPACE_RESOURCE; }
+  auto make_resource() -> resource* override
+  {
+    return new limited_memory_resource(mr_, allocation_limit_, alignment_);
+  }
 
  private:
-  rmm::mr::device_memory_resource* mr;
+  std::size_t allocation_limit_;
+  std::optional<std::size_t> alignment_;
+  std::shared_ptr<rmm::mr::device_memory_resource> mr_;
+
+  // Create a pool memory resource by default
+  static inline auto default_memory_resource(std::size_t limit)
```

Comment:
I like the idea of using the limiting memory resource so we can better compute batch sizes across raft (and downstream) algorithms. 

But if the user doesn't specify an explicit workspace memory resource to use, we should just default to wrapping a limited memory resource around the current default rmm resource and perhaps set the memory limit to the max memory of the device.

This allows a user to opt-in to this feature and gain all the benefit while not forcing us to make (potentially very bad) assumptions on the initial limits.  

This provides a very simple configuration option, allowing a user to specify "I want to use 3Gb for workspaces" while minimizing surprises from poorly set "automagical" settings that the uninformed or common user might now know anything about.

I think we should document this option heavily for the users, perhaps in a "getting the most out of RAFT" tutorial in the docs.

Reply:
In general, I agree having the default memory resource makes the most sense.
The only issue is if we want to use this in ivf-flat / ivf-pq, it will hit performance, because there we rely on the pooling allocator when creating many intermediate buffers on each iteration/batch when batching.

What do you think about this: query `rmm::mr::get_current_device_resource()` and wrap it in the pooling resource only if the resource is a plain `rmm::mr::cuda_memory_resource` (default rmm resource)?

Reply:
Alternatively, I could keep the default workspace resource plain/non-pooling unconditionally, but add a check to k-neighbours, ivf-pq, and ivf-flat with a warning sent to logs if the default workspace resource is used (suggesting to switch to the pooling workspace resource).

Reply:
Ok, done.